### PR TITLE
Allow to do serializing in machine-independent format

### PR DESCRIPTION
**Pull request by sorc1**
The patches introduce Pack()/Unpack() functions for Multi scanner to serialize/deserialize with msgpack.

### CODE REVIEW

File: pire/scanners/common.h [hunk 1]
```diff
@@ -61,6 +64,36 @@ namespace Pire {
 			if ((type != 0 && type != Type) || (hdrsize != 0 && HdrSize != hdrsize))
 				throw Error("Serialized regexp incompatible with your system");
 		}
+
+		void Pack(yostream* s)
+		{
+			char buf[PACKED_HEADERS * 8 * 4];
```
Comment (starius) [inline pire/scanners/common.h]: What is the meaning of 4? Could you replace 8 and 4 with named constants, please?

File: pire/scanners/multi.h [hunk 2]
```diff
@@ -592,6 +595,185 @@ struct ScannerSaver {
 		scanner.Swap(sc);
 	}
 
+	template<class Relocation, class Shortcutting>
+	static void PackScanner(const Scanner<Relocation, Shortcutting>& scanner, yostream* s)
+	{
+		Pire::Header hdr(1, 0);
+		char buf[256];
```
Comment (starius) [inline pire/scanners/multi.h]: What is the meaning of 256? Could you replace it with named constants, please?

File: pire/scanners/multi.h [hunk 3]
```diff
@@ -592,6 +595,185 @@ struct ScannerSaver {
...
+		char *ptr;
+		size_t i;
```
Comment (starius) [inline pire/scanners/multi.h]: Move this declaration to the loop where it is used.

File: pire/scanners/multi.h [hunk 4]
```diff
@@ -592,6 +595,185 @@ struct ScannerSaver {
 		scanner.Swap(sc);
 	}
 
+	template<class Relocation, class Shortcutting>
+	static void PackScanner(const Scanner<Relocation, Shortcutting>& scanner, yostream* s)
+	{
+		Pire::Header hdr(1, 0);
+		char buf[256];
+		char *ptr;
```
Comment (starius) [inline pire/scanners/multi.h]: Declare a variable at the same line with setting its initial value, please.

File: pire/scanners/multi.h [hunk 5]
```diff
@@ -592,6 +595,185 @@ struct ScannerSaver {
...
+		}
+	}
+
+	template<class Relocation, class Shortcutting>
+	static void UnpackScanner(Scanner<Relocation, Shortcutting>& scanner, yistream* s)
+	{
+		typedef Scanner<Relocation, Shortcutting> ScannerType;
+		Pire::Header hdr(1, 0);
+		Scanner<Relocation, Shortcutting> sc;
+		size_t i;
```
Comment (starius) [inline pire/scanners/multi.h]: Move to the line where it is used first time.

File: pire/scanners/multi.h [hunk 6]
```diff
@@ -773,6 +967,18 @@ class ExitMasks {
...
 	template <class Header>
+	static size_t GetMaskRaw(Header& header, size_t ind)
+	{
+		return header.Mask(ind);
+	}
+
+	template <class Header>
+	static void SetMaskRaw(Header& header, size_t ind, size_t val)
+	{
+		return header.SetMask(ind, val);
```
Comment (starius) [inline pire/scanners/multi.h]: s/return//

File: pire/scanners/multi.h [hunk 7]
```diff
@@ -773,6 +967,18 @@ class ExitMasks {
...
+	}
+
+	template <class Header>
 	static void SetMask(Header& header, size_t ind, char c)
 	{
 		header.SetMask(ind, FillSizeT(c));
```
Comment (starius) [inline pire/scanners/multi.h]: Use SetMaskRaw here.

File: pire/stub/defaults.h [hunk 8]
```diff
@@ -18,6 +18,12 @@
 #ifndef PIRE_STUB_DEFAULTS_H_INCLUDED
 #define PIRE_STUB_DEFAULTS_H_INCLUDED
 
+#if defined(__cplusplus) && !defined(__STDC_CONSTANT_MACROS)
+#define __STDC_CONSTANT_MACROS 1 /* make С++ to be happy */
+#endif
+#if defined(__cplusplus) && !defined(__STDC_LIMIT_MACROS)
+#define __STDC_LIMIT_MACROS 1    /* make С++ to be happy */
+#endif
```
Comment (starius) [inline pire/stub/defaults.h]: Explain this code more carefully, please.

File: pire/scanners/multi.h (hunk 7 above)
Comment (sorc1) [inline pire/scanners/multi.h]: Could you please explain what did you mean? What SetMaskRaw should I use here?

**Comment by sorc1** [comment]
21.09.2016 03:39, Boris Nagaev wrote:
>
> *@starius* requested changes on this pull request.
>
> General comments:
>
>  1. Style: add braces around one line blocks, please.
>  2. Why not protocol buffer?
>
Some reasons for that:
1. No extra build dependencies
2. The format is not going to change often
3. It's just simple: while packing/unpacking, we're working
     directly with the scanner instance's memory, not with 
objects-to-serialize

I'll fix the problems you've pointed on and then I'll resend the patches.

File: pire/scanners/multi.h (hunk 7 above)
Comment (starius) [inline pire/scanners/multi.h]: `SetMaskRaw<Header>(header, ind, FillSizeT(c));`

**Comment by sorc1** [comment]
#46

**Pull request closed without merging**